### DISCUSSION TITLE
How I Setting Custom Fonts | 我是如何设置自定义字体的

### DISCUSSION THREAD

**Discussion by zombie110year** [emilk/egui#373] [Show and tell]
> 我将这些代码写在了 [TemplateApp](https://github.com/emilk/egui_template/blob/341c11b9c7f24191f50dbfbd98f2488b1c387883/src/app.rs#L42) 的 `update` 方法里，因为我不知道在此之外的地方如何访问到 `CtxRef` 对象。如果有更合适的地方，请告知我。
>
> ---
>
> I wrote these codes into method `update` of [TemplateApp](https://github.com/emilk/egui_template/blob/341c11b9c7f24191f50dbfbd98f2488b1c387883/src/app.rs#L42)，because I didn't know how to acess `CtxRef` object outside the method. Please notice me if there were somewhere better to place these codes.

新用法示例（v0.18.0）：
The new example (v0.18.0)：

https://github.com/emilk/egui/blob/39b63f6163c7aabbf20138b505f0c86d62977408/examples/custom_font/src/main.rs

**Comment by zombie110year** [comment]
![图片](https://user-images.githubusercontent.com/36528430/117821891-98056780-b29e-11eb-8d89-e093ce465d58.png)

**Reply by emilk** [reply]
Yes, https://docs.rs/epi/0.12.0/epi/trait.App.html#method.setup is a better place!

**Reply by emilk** [reply]
You can also tell egui to first try your font, and then use the default fonts as fallback (for emojis and other characters not found in your font):

``` rust
// Put new font first (highest priority):
fonts.fonts_for_family.get_mut(&FontFamily::Monospace).unwrap().insert(0, "Sarasa".to_owned());
fonts.fonts_for_family.get_mut(&FontFamily::Proportional).unwrap().insert(0, "Sarasa".to_owned());
```

**Comment by logica-academy** [comment]
mutch thanks.
Just I needed topics.
I can now use Japanese as well.

    fn setup(&mut self, ctx: &egui::CtxRef){
        //Custom font install
        // # use epaint::text::*;
        // 1. Create a `FontDefinitions` object.
        let mut font = FontDefinitions::default();
        // Install my own font (maybe supporting non-latin characters):
        // 2. register the font content with a name.
        font.font_data.insert("mPlus".to_owned(),std::borrow::Cow::Borrowed(include_bytes!("../fonts/rounded-x-mplus-1p-bold.ttf")));
        //font.font_data.insert("mPlus".to_string(), Cow::from(&mPlus_font[..]));
        // 3. Set two font families to use the font, font's name must have been
        // Put new font first (highest priority)registered in `font_data`.
        font.fonts_for_family.get_mut(&FontFamily::Monospace).unwrap().insert(0, "mPlus".to_owned());
        font.fonts_for_family.get_mut(&FontFamily::Proportional).unwrap().insert(0, "mPlus".to_owned());
        // 4. Configure context with modified `FontDefinitions`.
        ctx.set_fonts(font);
    }

**Reply by zombie110year** [reply]
Maybe we should write a font loader to read fonts in runtime. My executable has 30+ MB by `include_bytes!`. 😆

**Reply by emilk** [reply]
You can load them at runtime and pass them to `std::borrow::Cow::Owned` instead. On native you can just read it from disk, on web you can use the http interface (`epi::http::Request::get(url)`, [missing in docs](https://github.com/emilk/egui/issues/381))

**Reply by moon989** [reply]
This is OK
`        // 设置字体；
        let mut font = egui::FontDefinitions::default();
        font.font_data.insert("Sarasa".to_owned(),std::borrow::Cow::Borrowed(include_bytes!("./fonts/sarasa-mono-sc-regular.woff2.ttf")));
        font.fonts_for_family.get_mut(&egui::FontFamily::Monospace).unwrap().insert(0, "Sarasa".to_owned());
        font.fonts_for_family.get_mut(&egui::FontFamily::Proportional).unwrap().insert(0, "Sarasa".to_owned());
        _ctx.set_fonts(font);`

**Reply by xuxiaowei-com-cn** [reply]
`Cargo.toml` 文件：

```
[package]
name = "cargo_egui"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
eframe = '0.18.0'

# eframe 内已引入 egui
# egui = '0.18.1'

```

`main.rs` 完整代码如下：

字体文件 `STSONG.TTF` 与 `main.rs` 在同一个文件夹

```rust
// 在发行版中隐藏 Windows 上的控制台窗口
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use eframe::egui;

fn main() {
    let options = eframe::NativeOptions::default();

    eframe::run_native(
        "我的 egui 应用",
        options,
        Box::new(|_cc| Box::new(People::default())),
    );
}

// 人
struct People {
    // 姓名
    name: String,
    // 年龄
    age: u32,
}

// 默认值
impl Default for People {
    fn default() -> Self {
        Self {
            name: "Arthur".to_owned(),
            age: 42,
        }
    }
}

//
impl eframe::App for People {
    // 每次 UI 需要重新绘制时调用
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {

        // 参考：https://github.com/emilk/egui/blob/0.17.0/eframe/examples/custom_font.rs

        // 此内容存在 bug，即：运行时闪烁，会从乱码转换到不乱码的状态

        // 从默认字体开始（我们将添加而不是替换它们）。
        let mut fonts = egui::FontDefinitions::default();

        // 安装我自己的字体（也许支持非拉丁字符）。
        // 支持 .ttf 和 .otf 文件。
        // STSONG.TTF 来自 Windows 10 系统的 C:\Windows\Fonts\STSONG.TTF
        fonts.font_data.insert("my_font".to_owned(), egui::FontData::from_static(include_bytes!("STSONG.TTF")));

        // 将我的字体放在首位（最高优先级）用于比例文本：
        fonts.families.entry(egui::FontFamily::Proportional).or_default().insert(0, "my_font".to_owned());

        // 将我的字体作为等宽字体的最后后备：
        fonts.families.entry(egui::FontFamily::Monospace).or_default().push("my_font".to_owned());

        // 告诉 egui 使用这些字体
        // 新字体将在下一帧开始时激活。
        // https://docs.rs/egui/latest/egui/struct.Context.html#method.set_fonts
        ctx.set_fonts(fonts);

        // 中央面板
        egui::CentralPanel::default().show(ctx, |ui| {

            // 显示大文本
            ui.heading("我的 egui 应用程序");

            // 使用水平布局启动 ui
            ui.horizontal(|ui| {
                // 显示一些文字
                ui.label("你的名字: ");
                // 不允许换行符，按下回车键将导致失去焦点
                ui.text_edit_singleline(&mut self.name);
            });

            // 滑块
            ui.add(egui::Slider::new(&mut self.age, 0..=120).text("岁"));

            // 按钮
            if ui.button("每年点击一次").clicked() {
                self.age += 1;
            }

            // 显示一些文字
            ui.label(format!("你好 '{}', 年龄 {}", self.name, self.age));
        });
    }
}

```

**Reply by zombie110year** [reply]
~~@xuxiaowei-com-cn 设置字体的那部分代码应该写在 `setup` 方法里，不要往 `update` 里面写。~~

旧 API 废弃了，看 [examples/custom_font](https://github.com/emilk/egui/blob/39b63f6163c7aabbf20138b505f0c86d62977408/examples/custom_font/src/main.rs) 里的用法。

**Reply by xuxiaowei-com-cn** [reply]
> @xuxiaowei-com-cn字体设置的那部分代码应该写在`setup`方法里，不要`update`往里面写。

不是很明白，请问该如何写？

**Reply by xuxiaowei-com-cn** [reply]
> @xuxiaowei-com-cn 设置字体的那部分代码应该写在 `setup` 方法里，不要往 `update` 里面写。

我并没有找到如何写 `setup`

**Reply by xuxiaowei-com-cn** [reply]
> @xuxiaowei-com-cn 设置字体的那部分代码应该写在 `setup` 方法里，不要往 `update` 里面写。

项目地址：https://github.com/xuxiaowei-com-cn/cargo_egui
请指点

**Comment by xuxiaowei-com-cn** [comment]
感谢 @zombie110year 提供的支持，
新文档：https://github.com/emilk/egui/blob/0.18.1/examples/custom_font/src/main.rs
上述代码修正的内容：https://github.com/xuxiaowei-com-cn/cargo_egui/pull/1

**Comment by AnlangA** [comment]
如果更改app标题的字体？例子里面的方式没法改变标题的字体
